barretenberg/cpp/src/barretenberg/circuit_checker/circuit_checker.cpp

```diff
@@ -132,27 +188,36 @@ void CircuitChecker::populate_values(
     values.w_l = builder.get_variable(block.w_l()[idx]);
     values.w_r = builder.get_variable(block.w_r()[idx]);
     values.w_o = builder.get_variable(block.w_o()[idx]);
-    if (memory_data.read_record_gates.contains(idx)) {
+    // Note: memory_data contains indices into the block to which RAM/ROM gates were added so we need to check that we
+    // are indexing into the correct block before updating the w_4 value.
+    if (memory_data.read_record_gates.contains(idx) && block.has_ram_rom) {
         values.w_4 = compute_memory_record_term(values.w_l, values.w_r, values.w_o, memory_data.eta);
-    } else if (memory_data.write_record_gates.contains(idx)) {
+    } else if (memory_data.write_record_gates.contains(idx) && block.has_ram_rom) {
         values.w_4 = compute_memory_record_term(values.w_l, values.w_r, values.w_o, memory_data.eta) + FF::one();
     } else {
         values.w_4 = builder.get_variable(block.w_4()[idx]);
     }
 
     // Set shifted wire values. Again, wire 4 is treated specially. On final row, set shift values to zero
-    values.w_l_shift = idx < block.size() - 1 ? builder.get_variable(block.w_l()[idx + 1]) : 0;
-    values.w_r_shift = idx < block.size() - 1 ? builder.get_variable(block.w_r()[idx + 1]) : 0;
-    values.w_o_shift = idx < block.size() - 1 ? builder.get_variable(block.w_o()[idx + 1]) : 0;
-    if (memory_data.read_record_gates.contains(idx + 1)) {
-        values.w_4_shift =
-            compute_memory_record_term(values.w_l_shift, values.w_r_shift, values.w_o_shift, memory_data.eta);
-    } else if (memory_data.write_record_gates.contains(idx + 1)) {
-        values.w_4_shift =
-            compute_memory_record_term(values.w_l_shift, values.w_r_shift, values.w_o_shift, memory_data.eta) +
-            FF::one();
+    if (idx < block.size() - 1) {
+        values.w_l_shift = builder.get_variable(block.w_l()[idx + 1]);
+        values.w_r_shift = builder.get_variable(block.w_r()[idx + 1]);
+        values.w_o_shift = builder.get_variable(block.w_o()[idx + 1]);
+        if (memory_data.read_record_gates.contains(idx + 1) && block.has_ram_rom) {
```

Comment:
It doesn't really matter, but block.has_ram_rom check is much faster than the contains check, so it's probably better to check it first in both cases

Reply:
good call